Performance Comparison: Data.table in R and Pandas in Python

DorisAmoakohene opened the pull request:
This is a draft on Performance Comparison: Data.table in R and Pandas in Python

DorisAmoakohene committed:
Create 2024-03-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene

DorisAmoakohene committed:
Delete posts/2024-03-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene

DorisAmoakohene committed:
Create index.qmd

DorisAmoakohene committed:
Update index.qmd

DorisAmoakohene commented:
@tdhock  and @Anirban166, Kindly help in Reviewing this Blog.

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+R and Python are two programming languages that have gained immense popularity among data scientists, statisticians, and researchers. 
+In this blog, we will explore two widely used libraries, data.table in R and pandas in Python, which excel in data manipulation and provide versatile functionalities for working with data, focusing on their capabilities for reading, writing, and reshaping data.
+We will also explain how to graphically demonstrate the time taken by each operation. We will use atime package to compare and visualize the asymptotic performance (time and memory usage) of the different functions mentioned above.
+By comparing the asymptotic performance of these packages in these programing languages, we aim to provide insights into their usage and help data scientists make informed choices when it comes to data manipulation and analysis.
```
good overall but please add an m in programing

DorisAmoakohene replied:
thanks, updated

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+
+In the atime::atime() we define the following arguments, 
+
+The first argument N is a sequence of data sizes.This explores computational efficiency, which will compute time/memory usage for different data sizes(N).
```
please add space after period

DorisAmoakohene replied:
done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+library(atime)
+library(ggplot2)
+library(reticulate)
+use_python("C:/Users/amoak/AppData/Local/Programs/Python/Python312/python.exe")
```
this will not be reproducible on another computer, so at least you should add a comment that says something like "If you want to reproduce, please change to the path of python on your computer"

DorisAmoakohene replied:
yes, thats the right thing to do

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+
+## Various plot
+
+atime::reference_best() : This plot shows the asymptotic reference lines, which defines the estimated asymptotic time and memory complexity. The output and figure will includes violet reference lines, in which the text labels can be interpreted in terms of big O notation (asymptotic time and memory usage).
```
add s references_best
and actually this function returns a list, which can then be visualized using plot()

tdhock replied:
also it would be better to move this text down below to where you actually use this function, and we can see the plot()

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+    input.vec <- rnorm(n.rows * N)
+    input.mat <- matrix(input.vec, n.rows, N)
+    input.df <- data.frame(input.mat)
+    
```
please delete empty lines

DorisAmoakohene replied:
done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,392 @@
+
+```{r,warning=FALSE,message=FALSE}
+refs.write.vary.cols <- atime::references_best(atime.write.vary.cols)
+plot(refs.write.vary.cols)
````
please include some commentary after every plot, including at least one sentence that tells us your interpretation of what the plot shows.

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,392 @@
+print(gg.write.dt.pd)
+```
+
+The Plot above shows that in terms of writing data, the data.table package in R outperforms Python's pandas library and particularly useful when dealing with large datasets.
````
this is good commentary

tdhock replied:
but please lower case Plot -> plot

DorisAmoakohene replied:
thanks,done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,392 @@
+plot(ml.wide2long.pd)
+
+```
+
````
space after comma please
lower case As -> as

DorisAmoakohene replied:
done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+
+# Conclusions 
+
+In conclusion, we have shown how to use atime to compare asymptotic time of the two packages. Both Pandas and Data.table are powerful libraries for data manipulation in Python and R respectively. Pandas offers a comprehensive set of functions and a user-friendly interface, making it suitable for a wide range of data analysis tasks. On the other hand, Data.table excels in terms of performance and  ...
```
Data.table -> `` `data.table` ``

DorisAmoakohene replied:
yes,done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,392 @@
+
+
+# References: 
+
```
please add at least one sentence here like "My code was copied and modified from the code linked below"

DorisAmoakohene replied:
done

DorisAmoakohene committed:
Update index.qmd

DorisAmoakohene committed:
Update index.qmd

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -88,6 +88,9 @@ plot(refs.write.vary.cols)
 atime::references_best() : This function returns a list, which can then be visualized using plot().
 The plot shows the asymptotic reference lines, which defines the estimated asymptotic time and memory complexity. The output and figure will includes violet reference lines, in which the text labels can be interpreted in terms of big O notation (asymptotic time and memory usage).
 
+The graph compares the performance of data.table::fwrite and pandas::to_csv for writing data. The top left graph shows that pandas::to_csv has a sharp increase in time as the dataset size increases, while data.table::fwrite remains relatively constant. The top right graph shows both methods with almost constant times, but pandas::to_csv uses more kilobytes. Overall, data.table::fwrite performs si ...
```
pandas::to_csv is potentially confusing because double colon (::) is R syntax, but you mean a python function/method right? Can you please use the python syntax?

DorisAmoakohene replied:
updated

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -159,7 +162,7 @@ atime.read <- atime::atime(
 plot(refs.read.vary.cols)
 ```
-
+The graph provides a closer look at larger values of N, indicating that data.table::fread outperforms pandas::read_csv, particularly as N increases. complexity notations such as log N, N log N, and N^2 are used to describe the performance characteristics.
````
capitalize C in complexity because it starts the sentence.
performance characteristics -> asymptotic time/memory usage

DorisAmoakohene replied:
updated

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+
+R and Python are two programming languages that have gained immense popularity among data scientists, statisticians, and researchers. 
+In this blog, we will explore two widely used libraries, data.table in R and pandas in Python, which excel in data manipulation and provide versatile functionalities for working with data, focusing on their capabilities for reading, writing, and reshaping data.
+We will also explain how to graphically demonstrate the time taken by each operation. We will use atime package to compare and visualize the asymptotic performance (time and memory usage) of the different functions mentioned above.
```
the* atime package

DorisAmoakohene replied:
done

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+R and Python are two programming languages that have gained immense popularity among data scientists, statisticians, and researchers. 
+In this blog, we will explore two widely used libraries, data.table in R and pandas in Python, which excel in data manipulation and provide versatile functionalities for working with data, focusing on their capabilities for reading, writing, and reshaping data.
+We will also explain how to graphically demonstrate the time taken by each operation. We will use atime package to compare and visualize the asymptotic performance (time and memory usage) of the different functions mentioned above.
+By comparing the asymptotic performance of these packages in these programming languages, we aim to provide insights into their usage and help data scientists make informed choices when it comes to data manipulation and analysis.
```
I* aim (since you're the one doing it and writing this, use first person singular)

tdhock replied:
we vs I -> either one is OK, just be consistent

DorisAmoakohene replied:
great

DorisAmoakohene replied:
I think I will go with the we, thanks

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+# Introduction:
+
+R and Python are two programming languages that have gained immense popularity among data scientists, statisticians, and researchers. 
+In this blog, we will explore two widely used libraries, data.table in R and pandas in Python, which excel in data manipulation and provide versatile functionalities for working with data, focusing on their capabilities for reading, writing, and reshaping data.
```
both of which

DorisAmoakohene replied:
Thanks

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+By comparing the asymptotic performance of these packages in these programming languages, we aim to provide insights into their usage and help data scientists make informed choices when it comes to data manipulation and analysis.
+
+
+In the atime::atime() we define the following arguments, 
```
In atime::atime()

Also code snippets, variable names, packages, etc. could do with formatting (like `atime::atime()` instead here)

DorisAmoakohene replied:
thanks

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+seconds.limit may optionally be specified. If an expression is slower than this limit for any data size, then no larger data sizes will be measured.
+
+Lastly, expr.list, which is the list of expressions for which time/memory usage will be measured.
+
```
The indentation overall (for this entire file) is inconsistent (no space, two spaces, one space) 

(like maybe just leave one line after each statement and stick with it?)

DorisAmoakohene replied:
addressed

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+library(atime)
+library(ggplot2)
+library(reticulate)
+use_python("C:/Users/amoak/AppData/Local/Programs/Python/Python312/python.exe") #If you want to reproduce, please change to the path of python on your computer.
```
Comments at the end of code is fine but for readability it drags the line out too much. Better to put the comment above the line where you call `use_python(" ... ")`

Anirban166 replied:
(especially when both the command and the comments are of considerable length)

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,372 @@
+plot(refs.write.vary.cols)
+```
+
+atime::references_best() : This function returns a list, which can then be visualized using plot().
````
There should be no gap/space between something and a colon

DorisAmoakohene replied:
done

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,372 @@
+print(gg.write.dt.pd)
+```
+
+The plot above shows that in terms of writing data, the data.table package in R outperforms Python's pandas library and particularly useful when dealing with large datasets.
````
I would rather say 'the pandas library in Python' since Pandas is not a part of Python (just like data.table is not a part of base R, and you mentioned 'the data.table package in R')

DorisAmoakohene replied:
yes, thanks

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+
+Data reshaping means changing the shape of the data, to get it into a more appropriate format, for learning/plotting/etc. Here we consider wide to long  and long to wide reshape, which means we start with a wide table (many columns) and end up with a long table (fewer columns) and vice versa. 
+
+## A.	wide to long reshape.
```
why is there so much space between 'A.' and 'wide'?

DorisAmoakohene replied:
It didnt show that way, to me, but thanks for drawing my attention to it, I have addressed it

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+
+On the other hand, when transforming data from long to wide format, data.table's dcast() function proves to be more efficient.
+
+Although data.table has advantages in terms of writing,reading and Rehaping, pandas in Python still provides a solid framework for data manipulation as seen in the plot.
```
writing, reading and reshaping

commas, typo checks, capitalization..check everything when you write please

DorisAmoakohene replied:
alright

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,372 @@
+    pd$pivot_table(py_df, values = "value", index = "id", columns = "category")
+  }
+  )
+
```
For code blocks too consistent indentation would be nice

like leave no extra line(s) at the end ideally (some blocks I see that, some I don't)

DorisAmoakohene replied:
alright

DorisAmoakohene committed:
Update index.qmd

DorisAmoakohene committed:
Update index.qmd

DorisAmoakohene commented:
@tdhock and @Anirban166 I have addressed all comments and I believe it this looks good for merge, you can go through and let me know if you think otherwise 
Thank you

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,339 @@
+atime::references_best(): This function returns a list, which can then be visualized using plot().
+The plot shows the asymptotic reference lines, which define the estimated asymptotic time and memory complexity. The output and figure will include violet reference lines, in which the text labels can be interpreted in terms of Big O notation (asymptotic time and memory usage).
+
+The graph compares the performance of data.table::fwrite and pandas.to_csv  for writing data. The top left graph shows that pandas.to_csv  has a sharp increase in time as the dataset size increases, while data.table::fwrite remains relatively constant. The top right graph shows both methods with almost constant times, but pandas.to_csv  uses more kilobytes. Overall, data.table::fwrite performs si ...
```
extra space between 'pandas.to_csv' and 'uses more kilobytes', and then before a comma (after 'pandas.to_csv ')

DorisAmoakohene replied:
gotchu

Anirban166 commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,339 @@
+plot(refs.read.vary.cols)
+```
+
+The graph provides a closer look at larger values of N, indicating that data.table::fread outperforms pandas.read_csv , particularly as N increases. Complexity notations such as log N, N log N, and N^2 are used to describe the asymptotic time/memory usage.
````
Extra space before comma again

DorisAmoakohene replied:
thanks

Anirban166 commented:
Should be good to go apart from a few minor things here and there (most likely Kelly will proofread and do final edits here as well)

If Toby agrees too, please convert it from draft to ready for review/merge.

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,339 @@
+We will also explain how to graphically demonstrate the time taken by each operation. We will use the atime package to compare and visualize the asymptotic performance (time and memory usage) of the different functions mentioned above.
+By comparing the asymptotic performance of these packages in these programming languages, We aim to provide insights into their usage and help data scientists make informed choices when it comes to data manipulation and analysis.
+
+In the ```atime::atime()``` we define the following arguments:
````
please use single backticks instead of triple for code on the same line as text

DorisAmoakohene replied:
thanks

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,339 @@
+library(reticulate)
+use_python("C:/Users/amoak/AppData/Local/Programs/Python/Python312/python.exe") #If you want to reproduce, please change to the path of python on your computer.
+virtualenv_create("fm-proj")
+use_virtualenv("fm-proj", required = F)
```
what is fm-proj?

DorisAmoakohene replied:
@tdhock, In the reticulate cheat sheet you have to create a virtual environment to help run pandas and help keep project dependencies separate also avoiding  conflicts between different projects. , and I named mine "fm_proj", so it just a given name to the virtual environment, and it can be any name  

https://ugoproto.github.io/ugo_r_doc/pdf/reticulate.pdf

tdhock replied:
you should explain that in your blog

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
````diff
@@ -0,0 +1,339 @@
+plot(refs.write.vary.cols)
+```
+
+atime::references_best(): This function returns a list, which can then be visualized using plot().
````
Why are you doing plot of references_best output? That is for comparing asymptotic complexity with references (linear, constant, quadratic) etc. In this case I guess pandas/data.table are both linear, right? In that case I don't think this plot is very interesting/informative, can you show us what it looks like?
probably the predict plot is more informative? (because it shows pandas/data.table on same axes, easier to see differences/ which one is more efficient)

DorisAmoakohene replied:
I will just take it out for now.
This is how the plot looks like for reading csv: 
![refs read vary cols](https://github.com/rdatatable-community/The-Raft/assets/143125519/c2a84c3a-b701-46a8-bdae-26b0dad3cc5f)

tdhock replied:
actually that is interesting, since pandas looks super-linear

DorisAmoakohene replied:
addressed

DorisAmoakohene replied:
this is for writing csv:
![refs write vary cols](https://github.com/rdatatable-community/The-Raft/assets/143125519/5bf08adc-328f-4ccd-a786-db67f40de073)

DorisAmoakohene committed:
Update index.qmd

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,324 @@
+The first argument, N, is a sequence of data sizes. This explores computational efficiency, which will compute time/memory usage for different data sizes (N). 
+The second argument setup is an expression that will be evaluated for each value in N, to create data of a given size.
+seconds.limit may optionally be specified. If an expression is slower than this limit for any data size, then no larger data sizes will be measured.
+Lastly, expr.list, which is the list of expressions for which time/memory usage will be measured.
```
`expr.list` is not used below so please clarify or delete

DorisAmoakohene replied:
done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,324 @@
+    data.table::fread("data.csv", showProgress = FALSE) 
+  },
+  "pandas.read_csv " = {
+    pd <- import("pandas")
+    reticulate::py_run_string("import pandas as pd")
```
this should be moved to setup if possible

DorisAmoakohene replied:
alright

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,324 @@
+    data.table::melt(data.table(df), id.vars = c("id",  "category"),variable.names="variable", value.name = "value")
+  },
+  "pd.melt" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
```
please move to setup if possible

DorisAmoakohene replied:
done

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,324 @@
+  "pd.melt" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
+    pd$melt(py_df, id_vars = c("id", "category"), value_name = "score")  
```
why is value_name="score" here but value.name="value" above?

DorisAmoakohene replied:
addressed

tdhock commented on posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd:
```diff
@@ -0,0 +1,324 @@
+    data.table::dcast(data.table(df), id ~ category, value.var = "value")
+  },
+  "pd$pivot_table" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
```
please move to setup if possible

DorisAmoakohene replied:
done

tdhock commented:
hi @DorisAmoakohene @Anirban166 can you please revise and then tell me when this is ready to review and merge?

DorisAmoakohene committed:
Update index.qmd

DorisAmoakohene commented:
I have revised all the comment, so I believe it good to go

tdhock commented:
it would make your blog post more convincing if you had some discussion of the overhead of running python code from reticulate in R. Is the overhead constant? (should be) Would be great if you could make an atime plot that shows that `py_run_string` has takes constant time  -- how much? 0.01, 0.001 seconds?

tdhock commented:
also please convert from draft to regular PR when you think it is ready for review.

DorisAmoakohene committed:
Update index.qmd

DorisAmoakohene commented:
@kbodwin Could you review this for merging. I have performed a  comparative benchmarking analysis, evaluating the performance of various functions in R's data.table and Python's pandas. This analysis provides valuable insights into the efficiency of these popular data manipulation libraries, highlighting their strengths and weaknesses. The benchmarking exercise compares the execution time and memory usage of equivalent functions in both libraries using the atime package in R.
The findings of this analysis will help inform decisions on which library to use for specific data manipulation tasks.

DorisAmoakohene commented:
> it would make your blog post more convincing if you had some discussion of the overhead of running python code from reticulate in R. Is the overhead constant? (should be) Would be great if you could make an atime plot that shows that `py_run_string` has takes constant time -- how much? 0.01, 0.001 seconds?

done

DorisAmoakohene commented:
> also please convert from draft to regular PR when you think it is ready for review.

okay

tdhock commented:
@kbodwin can you please publish this on the blog? I still don't see it.
